gossip/src/main.rs

```diff
@@ -142,13 +139,40 @@ fn main() -> Result<(), Box<dyn error::Error>> {
                 .value_of("node_pubkey")
                 .map(|pubkey_str| pubkey_str.parse::<Pubkey>().unwrap());
 
+            let mut gossip_addr = solana_netutil::parse_port_or_addr(
+                matches.value_of("gossip_port"),
+                socketaddr!(
+                    [127, 0, 0, 1],
+                    solana_netutil::find_available_port_in_range(VALIDATOR_PORT_RANGE)
+                        .expect("unable to find an available gossip port")
+                ),
+            );
+
+            // Run ip_echo_server so other nodes can join gossip through this node
+            let _ip_echo_server = solana_netutil::ip_echo_server(
+                std::net::TcpListener::bind(&gossip_addr).unwrap_or_else(|err| {
+                    eprintln!(
+                        "failed to bind ip echo server at tcp/{}: {}",
+                        gossip_addr, err
+                    );
+                    exit(1);
+                }),
+            );
+
+            gossip_addr.set_ip(
+                solana_netutil::get_public_ip_addr(&entrypoint_addr).unwrap_or_else(|err| {
```

Comment:
I don't understand this part. Why modify `gossip_addr` here?

Reply:
I think I understand now. Looks like `gossip_addr` is used to indicate to `discover(...)` that we want to join gossip. And `gossip_addr` is the entrypoint?

Reply:
I think it might be clearer if you use a separate var above like `my_gossip_addr` and then use its port for `gossip_addr.set_port(..)`

Reply:
Awesome, makes a lot more sense to me now!